### ISSUE TITLE
Order of attributes in generated HTML doesn't match order of attributes in lucid template

### ISSUE THREAD

**Issue by rashadg1030** [chrisdone/lucid#115]
Hello. I'm using lucid in conjunction with Alpine.js and htmx. These frameworks require that the order of attributes are specified in a specific order. Here's an example from my codebase:

```haskell
    button_
        -- TODO: Why do attributes render out of order?
        [ type_ "button"
        , xData_ $ if isToggled then Just "{ toggled: true }" else Just "{ toggled: false }"
        , class_ "relative inline-flex flex-shrink-0 ..."
        , role_ "switch"
        , if isDisabled then disabled_ "" else class_ ""
        , xOn_ "click" "toggled = !toggled"
        , xBind_ "class" "{ 'bg-indigo-600': toggled, 'bg-gray-200': !toggled }"
        ] $ do
                ...
```

renders as:

```html
<button
    data-x-on:click="toggled = !toggled"
    role="switch"
    data-x-bind:class="{ 'bg-indigo-600': toggled, 'bg-gray-200': !toggled }"
    data-x-data="{ toggled: false }"
    type="button"
    class="relative inline-flex flex-shrink-0 ..."
>
    ...
</button>
```

I need it to render as:

```html
<button
    type="button"
    data-x-data="{ toggled: false }"
    class="relative inline-flex flex-shrink-0 ..."
    role="switch"
    data-x-on:click="toggled = !toggled"
    data-x-bind:class="{ 'bg-indigo-600': toggled, 'bg-gray-200': !toggled }"
>
    ...
</button>
```

Is this because of something to do with `hashable`? Read some threads here on it but it didn't really make sense to me. Is there any way to fix this? Or is this just the way `lucid` is? Thank you for reading.

Edit: I'm using the latest version of `lucid` that's compatible with the latest version of `servant-lucid`. `lucid-2.9.12.1`.

**Comment by rashadg1030** [comment]
I see. So it seems to be fixed in `lucid-2.10.0` since it uses `Map` vs. `HashMap` now. So to fix my issue the version used by `servant-lucid` should be bumped to 2.10.0 and it should work fine?

**Comment by chrisdone** [comment]
The attributes are ordered by name, so using `Map` would make your attributes alphabetically ordered.

**Comment by rashadg1030** [comment]
> The attributes are ordered by name, so using `Map` would make your attributes alphabetically ordered.

Right! Of course. That makes sense. I should've caught that.... So, is there a way to get attributes to render in the specified order by any chance? It seems like I'm pretty much the only person that desires this feature, so is there anything I can do on my end to fix it for my use case? I like lucid a lot since it's pure Haskell and it allows me to use all the power of Haskell for my HTML templates, so I'd rather not switch to any other solution to be honest. Looking at blaze it seems like it doesn't rely on `Map` or `HashMap` so I assume it preserves the order of the attributes as they're defined? But I'll have to test it out. Not a fan of blaze syntax which is why I went with lucid in the first place. Hmmm.

**Comment by chrisdone** [comment]
I see. That’s fine. I think that while HTML attributes order shouldn’t matter in theory, it doesn’t hurt to preserve the original order as given. 

I think we can go ahead and use a list or Seq as the internal structure and then do a Map.fromListWith and zip [1..] over the attributes to merge keys.

That’ll give a stable order and merge duplicates.

**Comment by rashadg1030** [comment]
Okay great. So I tried to implement the possible solution but I don't know what I'm doing. First time actually contributing Lucid even though I'm a long time user. I tried the following based on what I understood from the comment you made above:

```haskell
newtype HtmlT m a =
  HtmlT {runHtmlT :: m (Map Int (Text, Text) -> Builder,a)
         -- ^ This is the low-level way to run the HTML transformer,
         -- finally returning an element builder and a value. You can
         -- pass 'mempty' for this argument for a top-level call. See
         -- 'evalHtmlT' and 'execHtmlT' for easier to use functions.
         }
```

I changed `Map Text Text` to `Map Int (Text, Text)` so ordering isn't based on the order of the attribute name. But this complicates the rest of the code excessively, so I'm assuming I'm going the wrong direction with this. I think I misunderstood your comment.

You mention using List or Seq as the "internal structure". Did you meant going from

```haskell
newtype HtmlT m a =
  HtmlT {runHtmlT :: m (Map Text Text -> Builder,a)
              }
```

to

```haskell
newtype HtmlT m a =
  HtmlT {runHtmlT :: m ([(Text, Text)] -> Builder,a)
              }
```
?

But then I thought that can't be the case because performance benefits from using Map would be lost, right?

I'm not sure but I'd like to help where I can.

I'm confused as to how the order of the attributes can be maintained with `Map Text Text` as the type, since you can only construct a Map of that type that's ordered by the Text key, which is the cause of the issue I'm experiencing.

**Comment by chrisdone** [comment]
> You mention using List or Seq as the "internal structure". Did you meant going from [..]

Yep. I meant e.g. importing `Seq` from `Data.Sequence`,

```haskell
newtype HtmlT m a =
  HtmlT {runHtmlT :: m (Seq (Text, Text) -> Builder,a)}
```

which supports [fast appending](https://hackage.haskell.org/package/containers-0.6.2.1/docs/Data-Sequence.html#v:-62--60-).

Then at [the time of rendering](https://github.com/chrisdone/lucid/blob/5718bbab42590ad61b4c4b7f52cd2c99a276823a/src/Lucid/Base.hs#L504-L533), just before producing output, you join duplicate attributes by concatenating the values.

```haskell
[("class","container"), ("id","foo"), ("class", " foo")]
```

Ought to yield e.g.

```html
<div class="container foo" id="foo">
```

This would mean a left-bias with respect to ordering. I.e. you choose the ordering first encountered on the left.

How to achieve that is up for debate.

**Comment by chrisdone** [comment]
Here's a reference implementation:

```haskell
> import Data.Containers.ListUtils
> import qualified Data.Map.Strict as M
> attributes = [("id","blob"),("class","container"), ("id","foo"), ("wibble","k"), ("class", " foo")]
> keys = nubOrd $ map fst attributes
> values = M.fromListWith (flip (<>)) attributes
> mapMaybe (\k -> fmap (k,) (M.lookup k values)) keys
[("id","blobfoo"),("class","container foo"),("wibble","k")]
```

**Comment by phadej** [comment]
Another major release. Yay!

**Comment by chrisdone** [comment]
Well, it would change the output of some people's tests (if anyone relied on the alphabetical ordering), so it'd be a breaking release anyway.

I don't feel particularly positive towards hidden or internal modules. Such things usually just get in my way as a library user. If someone's unhappy about a major release, I'll give them an understanding nod.

**Comment by phadej** [comment]
I meant internal as in `Lucid.Internal`, still public but could change in details.

The argument about non changing order is valid. You however should nail it precisely. Whether it's case-sensitive e.g. (HTML attributes aren't IIRC).

**Comment by chrisdone** [comment]
I feel like case insensitivity is a can of worms that we’d be better off not opening. Given that almost all attributes are used via a combinator, most people will never be affected by it. Even using the CI type, what constitutes case is complicated and counterintuitive. People will open bugs on this project because of some funny case, I am sure. But it’s a good point to raise, perhaps a doc string on the `Attribute` type would be fine? Something like, “attribute case isn’t normalized before merging or rendering.” CI in an HTML library feels like one of those “clever” additions that will not pay off more than it costs.

I think in the past I would go for an Internal module. But having used Internal modules from a dozen packages over the years, and also having had my code broken by ostensibly “external APIs”, I feel like the practice is lip service to those still married to Java-style encapsulation (there are some loud ones in the Haskell community), but in reality doesn’t change how people work. It just means as an author I have to move code around and do more indirection. I’m not opposed to it if you have the urge to do it because it would help you maintain the package better, though.

**Comment by rashadg1030** [comment]
Fixed by #117
